tsconfig.json

```
@@ -0,0 +1,15 @@
+{
+  "compilerOptions": {
+    "module": "esnext",
+    "target": "es2018",
```

Changing the target to be `es2018` would be a breaking change, as the current build produces an `es5` compatible build - albeit with a `.esm` version that uses the es6 export syntax. This is important because - while we don't use many new syntax features in this library - typically modules are not transformed outside of the scope of their import/export syntax, and so syntax like `const`  and `let` and the template string syntax will be left in.

It would be easier to transition without breaking changes by instead using `es5` here:

```suggestion
    "target": "es5
```

---

We need to target at least `es2017` so async/await and for…of syntax is not converted into large amounts of output code. Async/await support is the baseline, so downstream projects will need to include this module in their compiler configuration to support older runtimes.

---

Here's the diff between the codebase compiled with `es5` and `es2017`:

<details>

<summary>UMD</summary>

```
textarea-autosize:typescript* λ diff dist/umd/index.js ~/temp/es5-umd-index.js
13,17c13,17
<         let previousValue = null;
<         let isUserResized = false;
<         let x;
<         let y;
<         let height;
---
>         var previousValue = null;
>         var isUserResized = false;
>         var x;
>         var y;
>         var height;
20c20
<                 const newHeight = textarea.style.height;
---
>                 var newHeight = textarea.style.height;
31,32c31,32
<         const document = textarea.ownerDocument;
<         const documentElement = document.documentElement;
---
>         var document = textarea.ownerDocument;
>         var documentElement = document.documentElement;
34,35c34,35
<             let offsetTop = 0;
<             let el = textarea;
---
>             var offsetTop = 0;
>             var el = textarea;
40,42c40,42
<             const top = offsetTop - document.defaultView.pageYOffset;
<             const bottom = documentElement.clientHeight - (top + textarea.offsetHeight);
<             return { top, bottom };
---
>             var top = offsetTop - document.defaultView.pageYOffset;
>             var bottom = documentElement.clientHeight - (top + textarea.offsetHeight);
>             return { top: top, bottom: bottom };
51c51
<             const { top, bottom } = overflowOffset();
---
>             var _a = overflowOffset(), top = _a.top, bottom = _a.bottom;
55,57c55,57
<             const maxHeight = Number(getComputedStyle(textarea).height.replace(/px/, '')) + bottom;
<             textarea.style.maxHeight = `${maxHeight - 100}px`;
<             const container = textarea.parentElement;
---
>             var maxHeight = Number(getComputedStyle(textarea).height.replace(/px/, '')) + bottom;
>             textarea.style.maxHeight = maxHeight - 100 + "px";
>             var container = textarea.parentElement;
59c59
<                 const containerHeight = container.style.height;
---
>                 var containerHeight = container.style.height;
62c62
<                 textarea.style.height = `${textarea.scrollHeight}px`;
---
>                 textarea.style.height = textarea.scrollHeight + "px";
76c76
<         const form = textarea.form;
---
>         var form = textarea.form;
82c82
<             unsubscribe() {
---
>             unsubscribe: function () {
```

</details>

<details>

<summary>ESNext</summary>

```
textarea-autosize:typescript* λ diff dist/index.js ~/temp/es5-index.js
2,6c2,6
<     let previousValue = null;
<     let isUserResized = false;
<     let x;
<     let y;
<     let height;
---
>     var previousValue = null;
>     var isUserResized = false;
>     var x;
>     var y;
>     var height;
9c9
<             const newHeight = textarea.style.height;
---
>             var newHeight = textarea.style.height;
20,21c20,21
<     const document = textarea.ownerDocument;
<     const documentElement = document.documentElement;
---
>     var document = textarea.ownerDocument;
>     var documentElement = document.documentElement;
23,24c23,24
<         let offsetTop = 0;
<         let el = textarea;
---
>         var offsetTop = 0;
>         var el = textarea;
29,31c29,31
<         const top = offsetTop - document.defaultView.pageYOffset;
<         const bottom = documentElement.clientHeight - (top + textarea.offsetHeight);
<         return { top, bottom };
---
>         var top = offsetTop - document.defaultView.pageYOffset;
>         var bottom = documentElement.clientHeight - (top + textarea.offsetHeight);
>         return { top: top, bottom: bottom };
40c40
<         const { top, bottom } = overflowOffset();
---
>         var _a = overflowOffset(), top = _a.top, bottom = _a.bottom;
44,46c44,46
<         const maxHeight = Number(getComputedStyle(textarea).height.replace(/px/, '')) + bottom;
<         textarea.style.maxHeight = `${maxHeight - 100}px`;
<         const container = textarea.parentElement;
---
>         var maxHeight = Number(getComputedStyle(textarea).height.replace(/px/, '')) + bottom;
>         textarea.style.maxHeight = maxHeight - 100 + "px";
>         var container = textarea.parentElement;
48c48
<             const containerHeight = container.style.height;
---
>             var containerHeight = container.style.height;
51c51
<             textarea.style.height = `${textarea.scrollHeight}px`;
---
>             textarea.style.height = textarea.scrollHeight + "px";
65c65
<     const form = textarea.form;
---
>     var form = textarea.form;
71c71
<         unsubscribe() {
---
>         unsubscribe: function () {
```

</details>

There doesn't seem to be a huge difference between the two so I might just change this back to `es5` for now. `es2018` produced identical output to `es2017`.